# FINAL PROJECT - PART 1 
#### CLEAN HAPPINESS SCORES AROUND THE WORLD YEARS 2017/18/19/20

Data Sources:   
https://www.kaggle.com/unsdsn/world-happiness?select=2017.csv  
https://worldhappiness.report/ed/2020/

##### STEP 1 :
   - Get the datasets for 2017, 2018 and 2019 and merge them
   - Get the datasets from WHR 2020 to get the world region and average age in life  
    expectancy for each country and per year and merge it with the three main dataset. 

In [1]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
happy_2017=pd.read_csv('/Users/nuria/Documents/datasciencecourse/FINAL PROJECT/Final Prpoject Datasets/datasets_894_813759_2017.csv')
happy_2018=pd.read_csv('/Users/nuria/Documents/datasciencecourse/FINAL PROJECT/Final Prpoject Datasets/datasets_894_813759_2018.csv')
happy_2019=pd.read_csv('/Users/nuria/Documents/datasciencecourse/FINAL PROJECT/Final Prpoject Datasets/datasets_894_813759_2019.csv')


In [3]:
happy_WHR_region=pd.read_excel('/Users/nuria/Documents/datasciencecourse/FINAL PROJECT/Final Prpoject Datasets/WHR20_DataForFigure2.1.xls')
happy_WHR_age=pd.read_excel('/Users/nuria/Documents/datasciencecourse/FINAL PROJECT/Final Prpoject Datasets/WHR20_DataForTable2.1age.xls')

In [4]:
happy_2017.head(5)

Country  Happiness.Rank  Happiness.Score  Whisker.high  Whisker.low  \
0       Norway               1            7.537      7.594445     7.479556   
1      Denmark               2            7.522      7.581728     7.462272   
2      Iceland               3            7.504      7.622030     7.385970   
3  Switzerland               4            7.494      7.561772     7.426227   
4      Finland               5            7.469      7.527542     7.410458   

   Economy..GDP.per.Capita.    Family  Health..Life.Expectancy.   Freedom  \
0                  1.616463  1.533524                  0.796667  0.635423   
1                  1.482383  1.551122                  0.792566  0.626007   
2                  1.480633  1.610574                  0.833552  0.627163   
3                  1.564980  1.516912                  0.858131  0.620071   
4                  1.443572  1.540247                  0.809158  0.617951   

   Generosity  Trust..Government.Corruption.  Dystopia.Residual  
0    0.362012                       0.315964           2.277027  
1    0.355280                       0.400770           2.313707  
2    0.475540                       0.153527           2.322715  
3    0.290549                       0.367007           2.276716  
4    0.245483                       0.382612           2.430182

In [5]:
happy_2018.head(5)

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.632           1.305           1.592   
1             2            Norway  7.594           1.456           1.582   
2             3           Denmark  7.555           1.351           1.590   
3             4           Iceland  7.495           1.343           1.644   
4             5       Switzerland  7.487           1.420           1.549   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.874                         0.681       0.202   
1                    0.861                         0.686       0.286   
2                    0.868                         0.683       0.284   
3                    0.914                         0.677       0.353   
4                    0.927                         0.660       0.256   

   Perceptions of corruption  
0                      0.393  
1                      0.340  
2                      0.408  
3                      0.138  
4                      0.357

In [6]:
happy_2019.head(5)

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.769           1.340           1.587   
1             2           Denmark  7.600           1.383           1.573   
2             3            Norway  7.554           1.488           1.582   
3             4           Iceland  7.494           1.380           1.624   
4             5       Netherlands  7.488           1.396           1.522   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                    1.028                         0.603       0.271   
3                    1.026                         0.591       0.354   
4                    0.999                         0.557       0.322   

   Perceptions of corruption  
0                      0.393  
1                      0.410  
2                      0.341  
3                      0.118  
4                      0.298

In [7]:
happy_WHR_region.head(5)

Country name Regional indicator  Ladder score  \
0      Finland     Western Europe        7.8087   
1      Denmark     Western Europe        7.6456   
2  Switzerland     Western Europe        7.5599   
3      Iceland     Western Europe        7.5045   
4       Norway     Western Europe        7.4880   

   Standard error of ladder score  upperwhisker  lowerwhisker  \
0                        0.031156      7.869766      7.747634   
1                        0.033492      7.711245      7.579955   
2                        0.035014      7.628528      7.491272   
3                        0.059616      7.621347      7.387653   
4                        0.034837      7.556281      7.419719   

   Logged GDP per capita  Social support  Healthy life expectancy  \
0              10.639267        0.954330                71.900826   
1              10.774001        0.955991                72.402504   
2              10.979933        0.942847                74.102448   
3              10.772559        0.974670                73.000000   
4              11.087804        0.952487                73.200783   

   Freedom to make life choices  Generosity  Perceptions of corruption  \
0                      0.949172   -0.059482                   0.195445   
1                      0.951444    0.066202                   0.168489   
2                      0.921337    0.105911                   0.303728   
3                      0.948892    0.246944                   0.711710   
4                      0.955750    0.134533                   0.263218   

   Ladder score in Dystopia  Explained by: Log GDP per capita  \
0                  1.972317                          1.285190   
1                  1.972317                          1.326949   
2                  1.972317                          1.390774   
3                  1.972317                          1.326502   
4                  1.972317                          1.424207   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                      1.499526                               0.961271   
1                      1.503449                               0.979333   
2                      1.472403                               1.040533   
3                      1.547567                               1.000843   
4                      1.495173                               1.008072   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                    0.662317                  0.159670   
1                                    0.665040                  0.242793   
2                                    0.628954                  0.269056   
3                                    0.661981                  0.362330   
4                                    0.670201                  0.287985   

   Explained by: Perceptions of corruption  Dystopia + residual  
0                                 0.477857             2.762835  
1                                 0.495260             2.432741  
2                                 0.407946             2.350267  
3                                 0.144541             2.460688  
4                                 0.434101             2.168266

In [8]:
happy_WHR_age.head(5)

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0  Afghanistan  2008     3.723590            7.144916        0.450662   
1  Afghanistan  2009     4.401778            7.314788        0.552308   
2  Afghanistan  2010     4.758381            7.421525        0.539075   
3  Afghanistan  2011     3.831719            7.394349        0.521104   
4  Afghanistan  2012     3.782938            7.480296        0.520637   

   Healthy life expectancy at birth  Freedom to make life choices  Generosity  \
0                         50.799999                      0.718114    0.178993   
1                         51.200001                      0.678896    0.201228   
2                         51.599998                      0.600127    0.131578   
3                         51.919998                      0.495901    0.173452   
4                         52.240002                      0.530935    0.246943   

   Perceptions of corruption  Positive affect  ...  \
0                   0.881686         0.517637  ...   
1                   0.850035         0.583926  ...   
2                   0.706766         0.618265  ...   
3                   0.731109         0.611387  ...   
4                   0.775620         0.710385  ...   

   GINI index (World Bank estimate)  \
0                               NaN   
1                               NaN   
2                               NaN   
3                               NaN   
4                               NaN   

   GINI index (World Bank estimate), average 2000-2017, unbalanced panel  \
0                                                NaN                       
1                                                NaN                       
2                                                NaN                       
3                                                NaN                       
4                                                NaN                       

   gini of household income reported in Gallup, by wp5-year  \
0                                                NaN          
1                                           0.441906          
2                                           0.327318          
3                                           0.336764          
4                                           0.344540          

   Most people can be trusted, Gallup  \
0                                 NaN   
1                            0.286315   
2                            0.275833   
3                                 NaN   
4                                 NaN   

   Most people can be trusted, WVS round 1981-1984  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

   Most people can be trusted, WVS round 1989-1993  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

   Most people can be trusted, WVS round 1994-1998  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

   Most people can be trusted, WVS round 1999-2004  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

   Most people can be trusted, WVS round 2005-2009  \
0                                              NaN   
1                        

##### STEP 2:  
   - Select the columns I want to work with
   - From happy_WHR_region I will extract the scores for 2020
   - For the World Happiness Resources for 2017, 2018 and 2019 I will get the following 
    columns: Rank, Country, happynes overall rank,  health life expectancy score
   - I will rename all the columns of the 4 datasets to have the same names in the 
    columns

# 2017

In [9]:
happy_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 12 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        155 non-null    object 
 1   Happiness.Rank                 155 non-null    int64  
 2   Happiness.Score                155 non-null    float64
 3   Whisker.high                   155 non-null    float64
 4   Whisker.low                    155 non-null    float64
 5   Economy..GDP.per.Capita.       155 non-null    float64
 6   Family                         155 non-null    float64
 7   Health..Life.Expectancy.       155 non-null    float64
 8   Freedom                        155 non-null    float64
 9   Generosity                     155 non-null    float64
 10  Trust..Government.Corruption.  155 non-null    float64
 11  Dystopia.Residual              155 non-null    float64
dtypes: float64(10), int64(1), object(1)
memory usage: 

2017 Dataset :  We see that there are not null values. We have 155 entries, 12 columns.
1 variable is categorical and the other 11 variables are numerical

In [10]:
#get 2017 columns
#change the name of the columns I'm interested in
happy_2017.rename(columns={'Happiness.Rank':'Rank','Happiness.Score':'Happiness_score','Health..Life.Expectancy.':'Live_expect_score'}, errors="raise",inplace=True)
column_to_get=['Country','Rank','Happiness_score','Live_expect_score']
happy_2017_df=happy_2017[column_to_get]

In [11]:
happy_2017_df.head()

Country  Rank  Happiness_score  Live_expect_score
0       Norway     1            7.537           0.796667
1      Denmark     2            7.522           0.792566
2      Iceland     3            7.504           0.833552
3  Switzerland     4            7.494           0.858131
4      Finland     5            7.469           0.809158

# 2018

In [12]:
happy_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     155 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB


2018 Dataset: The first observation that we have to do is that in this dataset we have one more country than in 2017. We will have to check which one is. We haven't got null values. 

In [13]:
#change the name of the columns I'm interested in
happy_2018.rename(columns={'Overall rank':'Rank','Country or region':'Country','Score':'Happiness_score','Healthy life expectancy':'Live_expect_score'}, errors="raise",inplace=True)
column_to_get=['Country','Rank','Happiness_score','Live_expect_score']
happy_2018_df=happy_2018[column_to_get]

In [14]:
happy_2018_df.head(5)

Country  Rank  Happiness_score  Live_expect_score
0      Finland     1            7.632              0.874
1       Norway     2            7.594              0.861
2      Denmark     3            7.555              0.868
3      Iceland     4            7.495              0.914
4  Switzerland     5            7.487              0.927

In [15]:
happy_2018_df.Country.unique()

array(['Finland', 'Norway', 'Denmark', 'Iceland', 'Switzerland',
       'Netherlands', 'Canada', 'New Zealand', 'Sweden', 'Australia',
       'United Kingdom', 'Austria', 'Costa Rica', 'Ireland', 'Germany',
       'Belgium', 'Luxembourg', 'United States', 'Israel',
       'United Arab Emirates', 'Czech Republic', 'Malta', 'France',
       'Mexico', 'Chile', 'Taiwan', 'Panama', 'Brazil', 'Argentina',
       'Guatemala', 'Uruguay', 'Qatar', 'Saudi Arabia', 'Singapore',
       'Malaysia', 'Spain', 'Colombia', 'Trinidad & Tobago', 'Slovakia',
       'El Salvador', 'Nicaragua', 'Poland', 'Bahrain', 'Uzbekistan',
       'Kuwait', 'Thailand', 'Italy', 'Ecuador', 'Belize', 'Lithuania',
       'Slovenia', 'Romania', 'Latvia', 'Japan', 'Mauritius', 'Jamaica',
       'South Korea', 'Northern Cyprus', 'Russia', 'Kazakhstan', 'Cyprus',
       'Bolivia', 'Estonia', 'Paraguay', 'Peru', 'Kosovo', 'Moldova',
       'Turkmenistan', 'Hungary', 'Libya', 'Philippines', 'Honduras',
       'Belarus', 'Turkey

Let's try to discover why we have one more row in 2018 than in 2017:

In [16]:
countries17=happy_2017_df.Country.unique().tolist()

In [17]:
len(countries17)

155

In [18]:
countries18=happy_2018_df.Country.unique().tolist()

In [19]:
len(countries18)

156

In [20]:
happy_2018_df.Country.duplicated().sum()

0

In [21]:
# check what country is added in 2018 that in 2017 we did not have 

There are 4 countries that appear inm 2018 and looks like they are not in 2017
Let's check it deeper. The countires are: Taiwan Province of China, Trinidad and Tobago, North Cyprus, Hong Kong S.A.R, China

In [22]:
list_17=happy_2017_df.Country.sort_values(ascending=True)
list_18=happy_2018_df.Country.sort_values(ascending=True)

In [23]:
one_not_two = set(list_17).difference(list_18)

In [24]:
one_not_two

{'Hong Kong S.A.R., China',
 'North Cyprus',
 'Taiwan Province of China',
 'Trinidad and Tobago'}

In [25]:
two_not_one = set(list_18).difference(list_17)
two_not_one

{'Hong Kong', 'Laos', 'Northern Cyprus', 'Taiwan', 'Trinidad & Tobago'}

In [26]:
#change the name in 2017 for Hong Kong, North Cyprus , Taiwan and Trinidad & Tobago

We can see that 2017 don't have Laos in its study
I change the names of the countries in 2017 dataset that are written in a different way than 2018

In [27]:
count_to_rep={'Hong Kong S.A.R., China':'Hong Kong','North Cyprus':'Northern Cyprus',
              'Taiwan Province of China':'Taiwan','Trinidad and Tobago':'Trinidad & Tobago'}
happy_2017_df['Country']=happy_2017_df['Country'].replace(count_to_rep)

# 2019

In [28]:
happy_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     156 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB


2019 Dataset: We can see that we have 156 elements, 9 columns, and no null values. 
We have one categorical varible and 8 numerical variables

In [29]:
happy_2019.rename(columns={'Overall rank':'Rank','Country or region':'Country','Score':'Happiness_score','Healthy life expectancy':'Live_expect_score'}, errors="raise",inplace=True)
column_to_get=['Country','Rank','Happiness_score','Live_expect_score']
happy_2019_df=happy_2019[column_to_get]

In [30]:
happy_2019_df.head()

Country  Rank  Happiness_score  Live_expect_score
0      Finland     1            7.769              0.986
1      Denmark     2            7.600              0.996
2       Norway     3            7.554              1.028
3      Iceland     4            7.494              1.026
4  Netherlands     5            7.488              0.999

In [31]:
#let's check if there is any difference in the column names in 2019 and 2018 datasets

In [32]:
list_19=happy_2019_df.Country.sort_values(ascending=True)

In [33]:
one_not_two = set(list_18).difference(list_19)
one_not_two

{'Angola', 'Belize', 'Macedonia', 'Sudan'}

In [34]:
two_not_one = set(list_19).difference(list_18)
two_not_one

{'Comoros', 'Gambia', 'North Macedonia', 'Swaziland'}

In [35]:
#north macedonia:Macedonia 
happy_2019_df.Country.sort_values().unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Argentina', 'Armenia',
       'Australia', 'Austria', 'Azerbaijan', 'Bahrain', 'Bangladesh',
       'Belarus', 'Belgium', 'Benin', 'Bhutan', 'Bolivia',
       'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'Bulgaria',
       'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 'Canada',
       'Central African Republic', 'Chad', 'Chile', 'China', 'Colombia',
       'Comoros', 'Congo (Brazzaville)', 'Congo (Kinshasa)', 'Costa Rica',
       'Croatia', 'Cyprus', 'Czech Republic', 'Denmark',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador', 'Estonia',
       'Ethiopia', 'Finland', 'France', 'Gabon', 'Gambia', 'Georgia',
       'Germany', 'Ghana', 'Greece', 'Guatemala', 'Guinea', 'Haiti',
       'Honduras', 'Hong Kong', 'Hungary', 'Iceland', 'India',
       'Indonesia', 'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy',
       'Ivory Coast', 'Jamaica', 'Japan', 'Jordan', 'Kazakhstan', 'Kenya',
       'Kosovo', 'Kuwait', 'Kyrgyzstan', 'Lao

In [36]:
count_to_rep={'North Macedonia':'Macedonia'}
happy_2019_df['Country']=happy_2019_df['Country'].replace(count_to_rep)

In [37]:
happy_2019_df.Country.sort_values().unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Argentina', 'Armenia',
       'Australia', 'Austria', 'Azerbaijan', 'Bahrain', 'Bangladesh',
       'Belarus', 'Belgium', 'Benin', 'Bhutan', 'Bolivia',
       'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'Bulgaria',
       'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 'Canada',
       'Central African Republic', 'Chad', 'Chile', 'China', 'Colombia',
       'Comoros', 'Congo (Brazzaville)', 'Congo (Kinshasa)', 'Costa Rica',
       'Croatia', 'Cyprus', 'Czech Republic', 'Denmark',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador', 'Estonia',
       'Ethiopia', 'Finland', 'France', 'Gabon', 'Gambia', 'Georgia',
       'Germany', 'Ghana', 'Greece', 'Guatemala', 'Guinea', 'Haiti',
       'Honduras', 'Hong Kong', 'Hungary', 'Iceland', 'India',
       'Indonesia', 'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy',
       'Ivory Coast', 'Jamaica', 'Japan', 'Jordan', 'Kazakhstan', 'Kenya',
       'Kosovo', 'Kuwait', 'Kyrgyzstan', 'Lao

In [38]:
happy_2018_df.Country.sort_values().unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina',
       'Armenia', 'Australia', 'Austria', 'Azerbaijan', 'Bahrain',
       'Bangladesh', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Canada', 'Central African Republic', 'Chad', 'Chile', 'China',
       'Colombia', 'Congo (Brazzaville)', 'Congo (Kinshasa)',
       'Costa Rica', 'Croatia', 'Cyprus', 'Czech Republic', 'Denmark',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador', 'Estonia',
       'Ethiopia', 'Finland', 'France', 'Gabon', 'Georgia', 'Germany',
       'Ghana', 'Greece', 'Guatemala', 'Guinea', 'Haiti', 'Honduras',
       'Hong Kong', 'Hungary', 'Iceland', 'India', 'Indonesia', 'Iran',
       'Iraq', 'Ireland', 'Israel', 'Italy', 'Ivory Coast', 'Jamaica',
       'Japan', 'Jordan', 'Kazakhstan', 'Kenya', 'Kosovo', 'Kuwait',
       'Kyrgyzstan', 'Laos

In this case we see that there are some countries that appear in 2019 that in 2018 were not taken in consideration

In [39]:
happy_2019_df.head(1)

Country  Rank  Happiness_score  Live_expect_score
0  Finland     1            7.769              0.986

# 2020

In [40]:
happy_WHR_region.head(1)

Country name Regional indicator  Ladder score  \
0      Finland     Western Europe        7.8087   

   Standard error of ladder score  upperwhisker  lowerwhisker  \
0                        0.031156      7.869766      7.747634   

   Logged GDP per capita  Social support  Healthy life expectancy  \
0              10.639267         0.95433                71.900826   

   Freedom to make life choices  Generosity  Perceptions of corruption  \
0                      0.949172   -0.059482                   0.195445   

   Ladder score in Dystopia  Explained by: Log GDP per capita  \
0                  1.972317                           1.28519   

   Explained by: Social support  Explained by: Healthy life expectancy  \
0                      1.499526                               0.961271   

   Explained by: Freedom to make life choices  Explained by: Generosity  \
0                                    0.662317                   0.15967   

   Explained by: Perceptions of corruption  Dystopia + residual  
0                                 0.477857             2.762835

In [41]:
happy_WHR_region.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 20 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country name                                153 non-null    object 
 1   Regional indicator                          153 non-null    object 
 2   Ladder score                                153 non-null    float64
 3   Standard error of ladder score              153 non-null    float64
 4   upperwhisker                                153 non-null    float64
 5   lowerwhisker                                153 non-null    float64
 6   Logged GDP per capita                       153 non-null    float64
 7   Social support                              153 non-null    float64
 8   Healthy life expectancy                     153 non-null    float64
 9   Freedom to make life choices                153 non-null    float64
 10  Generosity    

In [42]:
#create the Rank columns because in the 2020 we don't have this column
happy_WHR_region['Rank']=range(1, 1+len(happy_WHR_region))

In [43]:
happy_WHR_region.Rank

0        1
1        2
2        3
3        4
4        5
      ... 
148    149
149    150
150    151
151    152
152    153
Name: Rank, Length: 153, dtype: int64

In [44]:
happy_WHR_region.rename(columns={'Country name':'Country','Ladder score':'Happiness_score',
                           'Explained by: Healthy life expectancy':'Live_expect_score',
                           'Country name':'Country','Healthy life expectancy':'Helth_life_exp_age'}, errors="raise",inplace=True)
columns_to_get=['Country','Rank','Happiness_score','Live_expect_score','Helth_life_exp_age']
happy_2020_df=happy_WHR_region[columns_to_get]

In [45]:
happy_2020_df

Country  Rank  Happiness_score  Live_expect_score  \
0                     Finland     1           7.8087           0.961271   
1                     Denmark     2           7.6456           0.979333   
2                 Switzerland     3           7.5599           1.040533   
3                     Iceland     4           7.5045           1.000843   
4                      Norway     5           7.4880           1.008072   
..                        ...   ...              ...                ...   
148  Central African Republic   149           3.4759           0.000000   
149                    Rwanda   150           3.3123           0.572383   
150                  Zimbabwe   151           3.2992           0.375038   
151               South Sudan   152           2.8166           0.208809   
152               Afghanistan   153           2.5669           0.266052   

     Helth_life_exp_age  
0             71.900826  
1             72.402504  
2             74.102448  
3             73.000000  
4             73.200783  
..                  ...  
148           45.200001  
149           61.098846  
150           55.617260  
151           51.000000  
152           52.590000  

[153 rows x 5 columns]

In [46]:
# Let's check if there is a Country written differently from 2018 and 2019

In [47]:
list_2018=happy_2018_df.Country.sort_values().unique()
list_2019=happy_2019_df.Country.sort_values().unique()
list_2020=happy_2020_df.Country.sort_values().unique()

In [48]:
one_not_two = set(list_2018).difference(list_2020)
one_not_two

{'Angola',
 'Belize',
 'Bhutan',
 'Hong Kong',
 'Northern Cyprus',
 'Qatar',
 'Somalia',
 'Sudan',
 'Syria',
 'Taiwan',
 'Trinidad & Tobago'}

In [49]:
two_not_one = set(list_2020).difference(list_2018)
two_not_one

{'Comoros',
 'Gambia',
 'Hong Kong S.A.R. of China',
 'Maldives',
 'North Cyprus',
 'Swaziland',
 'Taiwan Province of China',
 'Trinidad and Tobago'}

In [50]:
one_not_two = set(list_2019).difference(list_2020)
one_not_two

{'Bhutan',
 'Hong Kong',
 'Northern Cyprus',
 'Qatar',
 'Somalia',
 'Syria',
 'Taiwan',
 'Trinidad & Tobago'}

In [51]:
two_not_one = set(list_2020).difference(list_2019)
two_not_one

{'Hong Kong S.A.R. of China',
 'Maldives',
 'North Cyprus',
 'Taiwan Province of China',
 'Trinidad and Tobago'}

In [52]:
country_dic={'Hong Kong S.A.R. of China':'Hong Kong','North Cyprus':'Northern Cyprus',
             'Taiwan Province of China':'Taiwan','Trinidad and Tobago':'Trinidad & Tobago'}
happy_2020_df['Country']=happy_2020_df['Country'].replace(country_dic)

In [53]:
happy_2020_df.Country.sort_values().unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Argentina', 'Armenia',
       'Australia', 'Austria', 'Azerbaijan', 'Bahrain', 'Bangladesh',
       'Belarus', 'Belgium', 'Benin', 'Bolivia', 'Bosnia and Herzegovina',
       'Botswana', 'Brazil', 'Bulgaria', 'Burkina Faso', 'Burundi',
       'Cambodia', 'Cameroon', 'Canada', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros',
       'Congo (Brazzaville)', 'Congo (Kinshasa)', 'Costa Rica', 'Croatia',
       'Cyprus', 'Czech Republic', 'Denmark', 'Dominican Republic',
       'Ecuador', 'Egypt', 'El Salvador', 'Estonia', 'Ethiopia',
       'Finland', 'France', 'Gabon', 'Gambia', 'Georgia', 'Germany',
       'Ghana', 'Greece', 'Guatemala', 'Guinea', 'Haiti', 'Honduras',
       'Hong Kong', 'Hungary', 'Iceland', 'India', 'Indonesia', 'Iran',
       'Iraq', 'Ireland', 'Israel', 'Italy', 'Ivory Coast', 'Jamaica',
       'Japan', 'Jordan', 'Kazakhstan', 'Kenya', 'Kosovo', 'Kuwait',
       'Kyrgyzstan', 'Laos', 'Latvi

In [54]:
happy_2020_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Country             153 non-null    object 
 1   Rank                153 non-null    int64  
 2   Happiness_score     153 non-null    float64
 3   Live_expect_score   153 non-null    float64
 4   Helth_life_exp_age  153 non-null    float64
dtypes: float64(3), int64(1), object(1)
memory usage: 6.1+ KB


##### STEP 3:
   - Add a row in each dataset with its year, it will be useful in the plots
   - Add a column with the healthy life expectancy age in 2017, 2018 and 2019 datasets.

In [55]:
happy_2017_df['Year']=2017
happy_2018_df['Year']=2018
happy_2019_df['Year']=2019
happy_2020_df['Year']=2020

##### STEP 4:  
 - Merge the healthy life expectancy in the datasets 2017/2018/2019 because it does not  
   have it.

In [56]:
#Healthy life expectancy at birth , year
temp_2017=happy_WHR_age[happy_WHR_age.year==2017]
temp_2018=happy_WHR_age[happy_WHR_age.year==2018]
temp_2019=happy_WHR_age[happy_WHR_age.year==2019]

In [57]:
columns=['Country name','Healthy life expectancy at birth']
age_2017=temp_2017[columns]
age_2018=temp_2018[columns]
age_2019=temp_2019[columns]

In [58]:
#merge the three datasets for each year

In [59]:
#rename the columns
age_2017.rename(columns={'Country name':'Country','Healthy life expectancy at birth':'Helth_life_exp_age'}, errors="raise",inplace=True)
age_2018.rename(columns={'Country name':'Country','Healthy life expectancy at birth':'Helth_life_exp_age'}, errors="raise",inplace=True)
age_2019.rename(columns={'Country name':'Country','Healthy life expectancy at birth':'Helth_life_exp_age'}, errors="raise",inplace=True)

In [60]:
#first check that the countries are written the same way

In [61]:
A=age_2017.Country
B=happy_2017_df.Country

In [62]:
one_not_two = set(A).difference(B)
one_not_two

{'Gambia',
 'Hong Kong S.A.R. of China',
 'Laos',
 'Taiwan Province of China',
 'Trinidad and Tobago'}

In [63]:
two_not_one = set(B).difference(A)
two_not_one

{'Angola',
 'Belize',
 'Bhutan',
 'Burundi',
 'Hong Kong',
 'Malaysia',
 'Northern Cyprus',
 'Qatar',
 'Somalia',
 'Sudan',
 'Syria',
 'Taiwan',
 'Trinidad & Tobago'}

In [64]:
#Change the names of Hong Kong, Taiwan and Trinidad & Tobago

country_dic={'Hong Kong S.A.R. of China':'Hong Kong',
             'Taiwan Province of China':'Taiwan','Trinidad and Tobago':'Trinidad & Tobago'}
age_2017['Country']=age_2017['Country'].replace(country_dic)

In [65]:
#repeat the same procedures for 2018 and 2019

In [66]:
A=age_2018.Country
B=happy_2018_df.Country

In [67]:
one_not_two = set(A).difference(B)
one_not_two

{'Comoros',
 'Gambia',
 'Maldives',
 'North Cyprus',
 'Swaziland',
 'Taiwan Province of China'}

In [68]:
two_not_one = set(B).difference(A)
two_not_one

{'Angola',
 'Bahrain',
 'Belize',
 'Bhutan',
 'Central African Republic',
 'Congo (Kinshasa)',
 'Hong Kong',
 'Iceland',
 'Jamaica',
 'Kuwait',
 'Lesotho',
 'Northern Cyprus',
 'Paraguay',
 'Qatar',
 'Somalia',
 'South Sudan',
 'Sudan',
 'Syria',
 'Taiwan',
 'Trinidad & Tobago'}

In [69]:
country_dic={'Hong Kong S.A.R. of China':'Hong Kong','North Cyprus':'Northern Cyprus',
             'Taiwan Province of China':'Taiwan','Trinidad and Tobago':'Trinidad & Tobago'}
age_2018['Country']=age_2018['Country'].replace(country_dic)

In [70]:
A=age_2019.Country
B=happy_2019_df.Country

In [71]:
one_not_two = set(A).difference(B)
one_not_two

{'Hong Kong S.A.R. of China', 'North Cyprus', 'Taiwan Province of China'}

In [72]:
two_not_one = set(B).difference(A)
two_not_one

{'Armenia',
 'Bahrain',
 'Bhutan',
 'Burundi',
 'Central African Republic',
 'Congo (Kinshasa)',
 'Czech Republic',
 'Haiti',
 'Hong Kong',
 'Iraq',
 'Israel',
 'Jamaica',
 'Northern Cyprus',
 'Pakistan',
 'Qatar',
 'Russia',
 'Somalia',
 'South Sudan',
 'Syria',
 'Taiwan',
 'Trinidad & Tobago'}

In [73]:
country_dic={'Hong Kong S.A.R. of China':'Hong Kong','North Cyprus':'Northern Cyprus',
             'Taiwan Province of China':'Taiwan'}
age_2019['Country']=age_2019['Country'].replace(country_dic)

In [74]:
age_2019.Country.unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Argentina', 'Australia',
       'Austria', 'Azerbaijan', 'Bangladesh', 'Belarus', 'Belgium',
       'Benin', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Bulgaria', 'Burkina Faso', 'Cambodia', 'Cameroon', 'Canada',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros',
       'Congo (Brazzaville)', 'Costa Rica', 'Croatia', 'Cyprus',
       'Denmark', 'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador',
       'Estonia', 'Ethiopia', 'Finland', 'France', 'Gabon', 'Gambia',
       'Georgia', 'Germany', 'Ghana', 'Greece', 'Guatemala', 'Guinea',
       'Honduras', 'Hong Kong', 'Hungary', 'Iceland', 'India',
       'Indonesia', 'Iran', 'Ireland', 'Italy', 'Ivory Coast', 'Japan',
       'Jordan', 'Kazakhstan', 'Kenya', 'Kosovo', 'Kuwait', 'Kyrgyzstan',
       'Laos', 'Latvia', 'Lebanon', 'Lesotho', 'Liberia', 'Libya',
       'Lithuania', 'Luxembourg', 'Macedonia', 'Madagascar', 'Malawi',
       'Malaysia', 'Mali', 'Malta'

In [75]:
age_2017.reset_index(drop=True)

Country  Helth_life_exp_age
0    Afghanistan           52.799999
1        Albania           68.400002
2        Algeria           65.699997
3      Argentina           68.599998
4        Armenia           66.599998
..           ...                 ...
142    Venezuela           66.300003
143      Vietnam           67.699997
144        Yemen           55.900002
145       Zambia           54.799999
146     Zimbabwe           55.000000

[147 rows x 2 columns]

In [76]:
#Merge the three datasets
#
happy_2017_df=happy_2017_df.merge(age_2017, left_on='Country',right_on='Country',how='left')
happy_2018_df=happy_2018_df.merge(age_2018, left_on='Country',right_on='Country',how='left')
happy_2019_df=happy_2019_df.merge(age_2019, left_on='Country',right_on='Country',how='left')

In [77]:
happy_2017_df

Country  Rank  Happiness_score  Live_expect_score  Year  \
0                      Norway     1            7.537           0.796667  2017   
1                     Denmark     2            7.522           0.792566  2017   
2                     Iceland     3            7.504           0.833552  2017   
3                 Switzerland     4            7.494           0.858131  2017   
4                     Finland     5            7.469           0.809158  2017   
..                        ...   ...              ...                ...   ...   
150                    Rwanda   151            3.471           0.326425  2017   
151                     Syria   152            3.462           0.500533  2017   
152                  Tanzania   153            3.349           0.364509  2017   
153                   Burundi   154            2.905           0.151611  2017   
154  Central African Republic   155            2.693           0.018773  2017   

     Helth_life_exp_age  
0             73.099998  
1             72.099998  
2             73.000000  
3             73.800003  
4             71.800003  
..                  ...  
150           60.500000  
151                 NaN  
152           57.000000  
153                 NaN  
154           45.200001  

[155 rows x 6 columns]

In [78]:
happy_2018_df

Country  Rank  Happiness_score  Live_expect_score  Year  \
0                     Finland     1            7.632              0.874  2018   
1                      Norway     2            7.594              0.861  2018   
2                     Denmark     3            7.555              0.868  2018   
3                     Iceland     4            7.495              0.914  2018   
4                 Switzerland     5            7.487              0.927  2018   
..                        ...   ...              ...                ...   ...   
151                     Yemen   152            3.355              0.343  2018   
152                  Tanzania   153            3.303              0.381  2018   
153               South Sudan   154            3.254              0.177  2018   
154  Central African Republic   155            3.083              0.010  2018   
155                   Burundi   156            2.905              0.145  2018   

     Helth_life_exp_age  
0             71.900002  
1             73.199997  
2             72.400002  
3                   NaN  
4             74.099998  
..                  ...  
151           56.700001  
152           57.500000  
153                 NaN  
154                 NaN  
155           53.400002  

[156 rows x 6 columns]

##### STEP 5:
   - This next step is to get all the Datasets with the columns in the same order, to  
   concat them right.

In [79]:
happy_2019_df.head(1)

Country  Rank  Happiness_score  Live_expect_score  Year  Helth_life_exp_age
0  Finland     1            7.769              0.986  2019                72.0

In [80]:
happy_2020_df.head(1)

Country  Rank  Happiness_score  Live_expect_score  Helth_life_exp_age  Year
0  Finland     1           7.8087           0.961271           71.900826  2020

In [81]:
columns_order=['Year','Country','Rank','Happiness_score','Live_expect_score','Helth_life_exp_age']

In [82]:
happy_2017_df=happy_2017_df[columns_order]
happy_2018_df=happy_2018_df[columns_order]
happy_2019_df=happy_2019_df[columns_order]
happy_2020_df=happy_2020_df[columns_order]

In [83]:
happy_2017_df.head(1)

Year Country  Rank  Happiness_score  Live_expect_score  Helth_life_exp_age
0  2017  Norway     1            7.537           0.796667           73.099998

In [84]:
happy_2018_df.head(1)

Year  Country  Rank  Happiness_score  Live_expect_score  Helth_life_exp_age
0  2018  Finland     1            7.632              0.874           71.900002

In [85]:
happy_2019_df.head(1)

Year  Country  Rank  Happiness_score  Live_expect_score  Helth_life_exp_age
0  2019  Finland     1            7.769              0.986                72.0

In [86]:
happy_2020_df.head(1)

Year  Country  Rank  Happiness_score  Live_expect_score  Helth_life_exp_age
0  2020  Finland     1           7.8087           0.961271           71.900826

##### Step 6:
 - Merge the four datasets into one.  
 - Due to the merging noe I have some NAN values in 2017, 2018 and 2019 dataset in tha age column. I'm going to concat the four datasets and then I will substitute this null values by the average from the age in the other years for the same country. I have realized that in some datasets this info is missing but not in others

In [87]:
nulls2017=happy_2017_df[happy_2017_df['Helth_life_exp_age'].isnull()]
nulls2018=happy_2018_df[happy_2018_df['Helth_life_exp_age'].isnull()]
nulls2019=happy_2019_df[happy_2019_df['Helth_life_exp_age'].isnull()]

In [88]:
nulls2017

Year                  Country  Rank  Happiness_score  Live_expect_score  \
32   2017                   Taiwan    33            6.422           0.793984   
34   2017                    Qatar    35            6.375           0.710098   
41   2017                 Malaysia    42            6.084           0.618784   
49   2017                   Belize    50            5.956           0.450192   
60   2017          Northern Cyprus    61            5.810           0.834647   
70   2017                Hong Kong    71            5.472           0.943062   
77   2017                   Kosovo    78            5.279           0.541452   
92   2017                  Somalia    93            5.151           0.113989   
96   2017                   Bhutan    97            5.011           0.495879   
102  2017  Palestinian Territories   103            4.775           0.565667   
129  2017                    Sudan   130            4.139           0.290921   
139  2017                   Angola   140            3.795           0.049869   
151  2017                    Syria   152            3.462           0.500533   
153  2017                  Burundi   154            2.905           0.151611   

     Helth_life_exp_age  
32                  NaN  
34                  NaN  
41                  NaN  
49                  NaN  
60                  NaN  
70                  NaN  
77                  NaN  
92                  NaN  
96                  NaN  
102                 NaN  
129                 NaN  
139                 NaN  
151                 NaN  
153                 NaN

In [89]:
nulls2018

Year                   Country  Rank  Happiness_score  Live_expect_score  \
3    2018                   Iceland     4            7.495              0.914   
25   2018                    Taiwan    26            6.441              0.857   
31   2018                     Qatar    32            6.374              0.748   
37   2018         Trinidad & Tobago    38            6.192              0.564   
42   2018                   Bahrain    43            6.105              0.698   
44   2018                    Kuwait    45            6.083              0.675   
48   2018                    Belize    49            5.956              0.474   
55   2018                   Jamaica    56            5.890              0.693   
57   2018           Northern Cyprus    58            5.835              0.909   
63   2018                  Paraguay    64            5.681              0.615   
65   2018                    Kosovo    66            5.662              0.578   
75   2018                 Hong Kong    76            5.430              1.030   
96   2018                    Bhutan    97            5.082              0.527   
97   2018                   Somalia    98            4.982              0.115   
103  2018   Palestinian Territories   104            4.743              0.602   
131  2018          Congo (Kinshasa)   132            4.245              0.204   
136  2018                     Sudan   137            4.139              0.312   
140  2018                   Lesotho   141            3.808              0.079   
141  2018                    Angola   142            3.795              0.269   
149  2018                     Syria   150            3.462              0.539   
153  2018               South Sudan   154            3.254              0.177   
154  2018  Central African Republic   155            3.083              0.010   

     Helth_life_exp_age  
3                   NaN  
25                  NaN  
31                  NaN  
37                  NaN  
42                  NaN  
44                  NaN  
48                  NaN  
55                  NaN  
57                  NaN  
63                  NaN  
65                  NaN  
75                  NaN  
96                  NaN  
97                  NaN  
103                 NaN  
131                 NaN  
136                 NaN  
140                 NaN  
141                 NaN  
149                 NaN  
153                 NaN  
154                 NaN

In [90]:
nulls2019

Year                   Country  Rank  Happiness_score  Live_expect_score  \
12   2019                    Israel    13            7.139              1.029   
19   2019            Czech Republic    20            6.852              0.920   
24   2019                    Taiwan    25            6.446              0.914   
28   2019                     Qatar    29            6.374              0.871   
36   2019                   Bahrain    37            6.199              0.871   
38   2019         Trinidad & Tobago    39            6.192              0.713   
45   2019                    Kosovo    46            6.100              0.758   
55   2019                   Jamaica    56            5.890              0.831   
63   2019           Northern Cyprus    64            5.718              1.042   
66   2019                  Pakistan    67            5.653              0.535   
67   2019                    Russia    68            5.648              0.726   
75   2019                 Hong Kong    76            5.430              1.122   
94   2019                    Bhutan    95            5.082              0.604   
109  2019   Palestinian Territories   110            4.696              0.672   
111  2019                   Somalia   112            4.668              0.268   
115  2019                   Armenia   116            4.559              0.815   
125  2019                      Iraq   126            4.437              0.574   
126  2019          Congo (Kinshasa)   127            4.418              0.357   
144  2019                   Burundi   145            3.775              0.380   
146  2019                     Haiti   147            3.597              0.449   
148  2019                     Syria   149            3.462              0.440   
154  2019  Central African Republic   155            3.083              0.105   
155  2019               South Sudan   156            2.853              0.295   

     Helth_life_exp_age  
12                  NaN  
19                  NaN  
24                  NaN  
28                  NaN  
36                  NaN  
38                  NaN  
45                  NaN  
55                  NaN  
63                  NaN  
66                  NaN  
67                  NaN  
75                  NaN  
94                  NaN  
109                 NaN  
111                 NaN  
115                 NaN  
125                 NaN  
126                 NaN  
144                 NaN  
146                 NaN  
148                 NaN  
154                 NaN  
155                 NaN

In [91]:
list_to_concat=[happy_2017_df,happy_2018_df,happy_2019_df,happy_2020_df]
Happyness_df=pd.concat(list_to_concat)

In [92]:
Happyness_df=Happyness_df.reset_index(drop=True)

In [93]:
Happyness_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 620 entries, 0 to 619
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Year                620 non-null    int64  
 1   Country             620 non-null    object 
 2   Rank                620 non-null    int64  
 3   Happiness_score     620 non-null    float64
 4   Live_expect_score   620 non-null    float64
 5   Helth_life_exp_age  561 non-null    float64
dtypes: float64(3), int64(2), object(1)
memory usage: 29.2+ KB


In [94]:
Happyness_df.to_csv('Happyness_df_withNaN.csv')

##### STEP 7:
   - Take care of the null values in Health_life_exp_age in my final dataset Happyness

In [95]:
#get the countries
countries=Happyness_df.Country
type(countries)

pandas.core.series.Series

In [96]:
avg_age_country=pd.DataFrame(columns=['Country','Avg_age'])
avg_age_country

Empty DataFrame
Columns: [Country, Avg_age]
Index: []

In [97]:
#get the average of the Health_life_expectancy for each country
liiiiist=[]
for item in countries:
    pd_temp=Happyness_df[Happyness_df.Country==item]
    avg_age=pd_temp.Helth_life_exp_age.mean()
    listtt=[item,avg_age]
    liiiiist.append(listtt)
    superdf = pd.DataFrame(listtt,columns = ['ups'])
    superdf =superdf.T
    superdf.rename(columns={0:'Country',1:'Avg_age'}, inplace=True)
    superdf.reset_index(inplace=True)
    pd.concat([avg_age_country, superdf], ignore_index=True)


In [98]:
liiiiist

[['Norway', 73.2001953125],
 ['Denmark', 72.40062522888184],
 ['Iceland', 73.0],
 ['Switzerland', 74.10061264038086],
 ['Finland', 71.90020751953125],
 ['Netherlands', 72.30023002624512],
 ['Canada', 73.60040092468262],
 ['New Zealand', 73.20065689086914],
 ['Sweden', 72.60019111633301],
 ['Australia', 73.60113525390625],
 ['Israel', 73.20008595784505],
 ['Costa Rica', 71.29996299743652],
 ['Austria', 73.00062561035156],
 ['United States', 68.29987525939941],
 ['Ireland', 72.30019760131836],
 ['Germany', 72.20050430297852],
 ['Belgium', 72.00041198730469],
 ['Luxembourg', 72.5999984741211],
 ['United Kingdom', 72.30040168762207],
 ['Chile', 69.90015602111816],
 ['United Arab Emirates', 67.0956974029541],
 ['Brazil', 66.42004013061523],
 ['Czech Republic', 70.04931386311848],
 ['Argentina', 68.80095100402832],
 ['Mexico', 68.29975700378418],
 ['Singapore', 76.80114555358887],
 ['Malta', 72.19999694824219],
 ['Uruguay', 69.00063705444336],
 ['Guatemala', 64.80223655700684],
 ['Panama', 6

In [99]:
Contry_Ageavg = pd.DataFrame(liiiiist, columns=['Country','Age_avg'])

In [100]:
Contry_Ageavg

Country    Age_avg
0                      Norway  73.200195
1                     Denmark  72.400625
2                     Iceland  73.000000
3                 Switzerland  74.100613
4                     Finland  71.900208
..                        ...        ...
615  Central African Republic  45.200001
616                    Rwanda  61.099711
617                  Zimbabwe  55.604315
618               South Sudan  51.000000
619               Afghanistan  52.597500

[620 rows x 2 columns]

- To replace the NaN values in my final dataset I replace the NaN for 0 and then I merge the datset with the countries and age average. Once I have the merge dones I replace the 0 values for the average. 

In [101]:
Happyness_clean_df=Happyness_df.merge(Contry_Ageavg, left_on='Country',right_on='Country',how='left')

In [102]:
Happyness_clean_df.drop_duplicates(inplace=True)

In [103]:
Happyness_clean_df.fillna(value=0, inplace=True)

In [104]:
Happyness_clean_df.loc[Happyness_clean_df['Helth_life_exp_age']==0,'Helth_life_exp_age']=Happyness_clean_df.Age_avg

In [105]:
Happyness_clean_df[Happyness_clean_df.Country=='Taiwan']

Year Country  Rank  Happiness_score  Live_expect_score  \
128   2017  Taiwan    33           6.4220           0.793984   
710   2018  Taiwan    26           6.4410           0.857000   
1319  2019  Taiwan    25           6.4460           0.914000   
1932  2020  Taiwan    25           6.4554           0.878438   

      Helth_life_exp_age    Age_avg  
128            69.599998  69.599998  
710            69.599998  69.599998  
1319           69.599998  69.599998  
1932           69.599998  69.599998

In [106]:
Happyness_clean_df[Happyness_clean_df.Country=='Malaysia']

Year   Country  Rank  Happiness_score  Live_expect_score  \
163   2017  Malaysia    42           6.0840           0.618784   
745   2018  Malaysia    35           6.3220           0.669000   
1538  2019  Malaysia    80           5.3390           0.828000   
2160  2020  Malaysia    82           5.3843           0.788512   

      Helth_life_exp_age    Age_avg  
163            67.100718  67.100718  
745            67.000000  67.100718  
1538           67.199997  67.100718  
2160           67.102158  67.100718

In [107]:
Happyness_clean_df[Happyness_clean_df.Helth_life_exp_age==0]

Year  Country  Rank  Happiness_score  Live_expect_score  \
136   2017    Qatar    35            6.375           0.710098   
195   2017   Belize    50            5.956           0.450192   
365   2017  Somalia    93            5.151           0.113989   
380   2017   Bhutan    97            5.011           0.495879   
511   2017    Sudan   130            4.139           0.290921   
549   2017   Angola   140            3.795           0.049869   
595   2017    Syria   152            3.462           0.500533   
734   2018    Qatar    32            6.374           0.748000   
801   2018   Belize    49            5.956           0.474000   
991   2018   Bhutan    97            5.082           0.527000   
994   2018  Somalia    98            4.982           0.115000   
1148  2018    Sudan   137            4.139           0.312000   
1166  2018   Angola   142            3.795           0.269000   
1196  2018    Syria   150            3.462           0.539000   
1335  2019    Qatar    29            6.374           0.871000   
1598  2019   Bhutan    95            5.082           0.604000   
1664  2019  Somalia   112            4.668           0.268000   
1805  2019    Syria   149            3.462           0.440000   

      Helth_life_exp_age  Age_avg  
136                  0.0      0.0  
195                  0.0      0.0  
365                  0.0      0.0  
380                  0.0      0.0  
511                  0.0      0.0  
549                  0.0      0.0  
595                  0.0      0.0  
734                  0.0      0.0  
801                  0.0      0.0  
991                  0.0      0.0  
994                  0.0      0.0  
1148                 0.0      0.0  
1166                 0.0      0.0  
1196                 0.0      0.0  
1335                 0.0      0.0  
1598                 0.0      0.0  
1664                 0.0      0.0  
1805                 0.0      0.0

In [108]:
Happyness_clean_df[(Happyness_clean_df.Country=='Somalia')&((Happyness_clean_df.Year==2020))]

Empty DataFrame
Columns: [Year, Country, Rank, Happiness_score, Live_expect_score, Helth_life_exp_age, Age_avg]
Index: []

In [109]:
Happyness_clean_df.to_csv('Happyness_clean_df_with_zeros.csv')

   - Because there are values in the 'Helth_life_exp_age' column but I don't hace the info from the WHR2020 because these countries where not in that study I get them searching online. Tha life expectancy for these countries are obtained from:  
   https://www.macrotrends.net/countries/QAT/qatar/life-expectancy

In [110]:
Happyness_cleaned_nozeros=pd.read_csv('/Users/nuria/Documents/datasciencecourse/FINAL PROJECT/Happyness_clean_df_nozero.csv')

In [111]:
Happyness_cleaned_nozeros.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 620 entries, 0 to 619
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Unnamed: 0          620 non-null    int64  
 1   Year                620 non-null    int64  
 2   Country             620 non-null    object 
 3   Rank                620 non-null    int64  
 4   Happiness_score     620 non-null    float64
 5   Live_expect_score   620 non-null    float64
 6   Helth_life_exp_age  620 non-null    float64
 7   Age_avg             620 non-null    float64
dtypes: float64(4), int64(3), object(1)
memory usage: 38.9+ KB


In [112]:
Happyness_cleaned_nozeros

Unnamed: 0  Year                   Country  Rank  Happiness_score  \
0             0  2017                    Norway     1           7.5370   
1             4  2017                   Denmark     2           7.5220   
2             8  2017                   Iceland     3           7.5040   
3            12  2017               Switzerland     4           7.4940   
4            16  2017                   Finland     5           7.4690   
..          ...   ...                       ...   ...              ...   
615        2418  2020  Central African Republic   149           3.4759   
616        2422  2020                    Rwanda   150           3.3123   
617        2426  2020                  Zimbabwe   151           3.2992   
618        2430  2020               South Sudan   152           2.8166   
619        2434  2020               Afghanistan   153           2.5669   

     Live_expect_score  Helth_life_exp_age    Age_avg  
0             0.796667           73.099998  73.200195  
1             0.792566           72.099998  72.400625  
2             0.833552           73.000000  73.000000  
3             0.858131           73.800003  74.100613  
4             0.809158           71.800003  71.900208  
..                 ...                 ...        ...  
615           0.000000           45.200001  45.200001  
616           0.572383           61.098846  61.099711  
617           0.375038           55.617260  55.604315  
618           0.208809           51.000000  51.000000  
619           0.266052           52.590000  52.597500  

[620 rows x 8 columns]

In [113]:
Happyness_cleaned_nozeros.drop(['Unnamed: 0','Age_avg'], axis=1,inplace=True)

In [114]:
Happyness_cleaned_nozeros[Happyness_cleaned_nozeros.Country=='Angola']

Year Country  Rank  Happiness_score  Live_expect_score  \
139  2017  Angola   140            3.795           0.049869   
296  2018  Angola   142            3.795           0.269000   

     Helth_life_exp_age  
139               59.98  
296               60.54

In [115]:
Happyness_cleaned_nozeros.to_csv('Happyness cleaned and age average.csv')

##### Setp 8:
   - Merge the Region Column in the dataset

In [116]:
#get the region and the country from happy_WHR_region

In [117]:
columns=['Country','Regional indicator']
regions_df=happy_WHR_region[columns]

In [118]:
regions_df

Country  Regional indicator
0                     Finland      Western Europe
1                     Denmark      Western Europe
2                 Switzerland      Western Europe
3                     Iceland      Western Europe
4                      Norway      Western Europe
..                        ...                 ...
148  Central African Republic  Sub-Saharan Africa
149                    Rwanda  Sub-Saharan Africa
150                  Zimbabwe  Sub-Saharan Africa
151               South Sudan  Sub-Saharan Africa
152               Afghanistan          South Asia

[153 rows x 2 columns]

In [119]:
Happyness_in_world=Happyness_cleaned_nozeros.merge(regions_df, left_on='Country',right_on='Country',how='left')

In [120]:
Happyness_in_world

Year                   Country  Rank  Happiness_score  Live_expect_score  \
0    2017                    Norway     1           7.5370           0.796667   
1    2017                   Denmark     2           7.5220           0.792566   
2    2017                   Iceland     3           7.5040           0.833552   
3    2017               Switzerland     4           7.4940           0.858131   
4    2017                   Finland     5           7.4690           0.809158   
..    ...                       ...   ...              ...                ...   
615  2020  Central African Republic   149           3.4759           0.000000   
616  2020                    Rwanda   150           3.3123           0.572383   
617  2020                  Zimbabwe   151           3.2992           0.375038   
618  2020               South Sudan   152           2.8166           0.208809   
619  2020               Afghanistan   153           2.5669           0.266052   

     Helth_life_exp_age  Regional indicator  
0             73.099998      Western Europe  
1             72.099998      Western Europe  
2             73.000000      Western Europe  
3             73.800003      Western Europe  
4             71.800003      Western Europe  
..                  ...                 ...  
615           45.200001  Sub-Saharan Africa  
616           61.098846  Sub-Saharan Africa  
617           55.617260  Sub-Saharan Africa  
618           51.000000  Sub-Saharan Africa  
619           52.590000          South Asia  

[620 rows x 7 columns]

In [123]:
#get the uniques Regional Indicators that we have
Happyness_in_world['Regional indicator'].unique()

array(['Western Europe', 'North America and ANZ',
       'Middle East and North Africa', 'Latin America and Caribbean',
       'Central and Eastern Europe', 'Southeast Asia', nan,
       'Commonwealth of Independent States', 'East Asia',
       'Sub-Saharan Africa', 'South Asia'], dtype=object)

In [124]:
#rename the name of the columns
Happyness_in_world.rename(columns={"Regional indicator": "Region"},inplace=True)

In [125]:
Happyness_in_world.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 620 entries, 0 to 619
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Year                620 non-null    int64  
 1   Country             620 non-null    object 
 2   Rank                620 non-null    int64  
 3   Happiness_score     620 non-null    float64
 4   Live_expect_score   620 non-null    float64
 5   Helth_life_exp_age  620 non-null    float64
 6   Region              586 non-null    object 
dtypes: float64(3), int64(2), object(2)
memory usage: 38.8+ KB


In [126]:
#get the countries that we have null values
d=Happyness_in_world[Happyness_in_world.Region.isnull()]

In [127]:
d #These are the countries without a Region assigned

Year            Country  Rank  Happiness_score  Live_expect_score  \
32   2017             Taiwan    33           6.4220           0.793984   
34   2017              Qatar    35           6.3750           0.710098   
37   2017  Trinidad & Tobago    38           6.1680           0.519983   
49   2017             Belize    50           5.9560           0.450192   
60   2017    Northern Cyprus    61           5.8100           0.834647   
70   2017          Hong Kong    71           5.4720           0.943062   
92   2017            Somalia    93           5.1510           0.113989   
96   2017             Bhutan    97           5.0110           0.495879   
129  2017              Sudan   130           4.1390           0.290921   
139  2017             Angola   140           3.7950           0.049869   
151  2017              Syria   152           3.4620           0.500533   
180  2018             Taiwan    26           6.4410           0.857000   
186  2018              Qatar    32           6.3740           0.748000   
192  2018  Trinidad & Tobago    38           6.1920           0.564000   
203  2018             Belize    49           5.9560           0.474000   
212  2018    Northern Cyprus    58           5.8350           0.909000   
230  2018          Hong Kong    76           5.4300           1.030000   
251  2018             Bhutan    97           5.0820           0.527000   
252  2018            Somalia    98           4.9820           0.115000   
291  2018              Sudan   137           4.1390           0.312000   
296  2018             Angola   142           3.7950           0.269000   
304  2018              Syria   150           3.4620           0.539000   
335  2019             Taiwan    25           6.4460           0.914000   
339  2019              Qatar    29           6.3740           0.871000   
349  2019  Trinidad & Tobago    39           6.1920           0.713000   
374  2019    Northern Cyprus    64           5.7180           1.042000   
386  2019          Hong Kong    76           5.4300           1.122000   
405  2019             Bhutan    95           5.0820           0.604000   
422  2019            Somalia   112           4.6680           0.268000   
459  2019              Syria   149           3.4620           0.440000   
491  2020             Taiwan    25           6.4554           0.878438   
508  2020  Trinidad & Tobago    42           6.1919           0.658829   
542  2020    Northern Cyprus    76           5.5355           1.026125   
544  2020          Hong Kong    78           5.5104           1.136631   

     Helth_life_exp_age Region  
32            69.599998    NaN  
34            79.910000    NaN  
37            63.500000    NaN  
49            74.180000    NaN  
60            73.702225    NaN  
70            76.771706    NaN  
92            56.520000    NaN  
96            70.910000    NaN  
129           64.710000    NaN  
139           59.980000    NaN  
151           70.970000    NaN  
180           69.599998    NaN  
186           80.020000    NaN  
192           63.500000    NaN  
203           74.370000    NaN  
212           73.702225    NaN  
230           76.771706    NaN  
251           71.280000    NaN  
252           56.940000    NaN  
291           64.950000    NaN  
296           60.540000    NaN  
304           71.780000    NaN  
335           69.599998    NaN  
339           80.160000    NaN  
349           63.500000    NaN  
374           73.702225    NaN  
386           76.771706    NaN  
405           71.580000    NaN  
422           57.220000    NaN  
459           72.700000    NaN  
491           69.599998    NaN  
508           63.500000    NaN  
542           73.702225    NaN  
544           76.771706    NaN

In [128]:
Happyness_in_world[Happyness_in_world.Region=='Sub-Saharan Africa']

Year                   Country  Rank  Happiness_score  Live_expect_score  \
63   2017                 Mauritius    64           5.6290           0.638007   
94   2017                   Nigeria    95           5.0740           0.056916   
100  2017              South Africa   101           4.8290           0.187080   
105  2017              Sierra Leone   106           4.7090           0.005565   
106  2017                  Cameroon   107           4.6950           0.132892   
..    ...                       ...   ...              ...                ...   
614  2020                  Tanzania   148           3.4762           0.442678   
615  2020  Central African Republic   149           3.4759           0.000000   
616  2020                    Rwanda   150           3.3123           0.572383   
617  2020                  Zimbabwe   151           3.2992           0.375038   
618  2020               South Sudan   152           2.8166           0.208809   

     Helth_life_exp_age              Region  
63            66.099998  Sub-Saharan Africa  
94            49.299999  Sub-Saharan Africa  
100           56.099998  Sub-Saharan Africa  
105           49.200001  Sub-Saharan Africa  
106           51.900002  Sub-Saharan Africa  
..                  ...                 ...  
614           57.496075  Sub-Saharan Africa  
615           45.200001  Sub-Saharan Africa  
616           61.098846  Sub-Saharan Africa  
617           55.617260  Sub-Saharan Africa  
618           51.000000  Sub-Saharan Africa  

[150 rows x 7 columns]

In [129]:
Happyness_in_world[Happyness_in_world.Country=='Iraq']

Year Country  Rank  Happiness_score  Live_expect_score  \
116  2017    Iraq   117           4.4970           0.501180   
271  2018    Iraq   117           4.4560           0.536000   
436  2019    Iraq   126           4.4370           0.574000   
576  2020    Iraq   110           4.7848           0.529351   

     Helth_life_exp_age                        Region  
116           59.599998  Middle East and North Africa  
271           60.200001  Middle East and North Africa  
436           59.901183  Middle East and North Africa  
576           59.903549  Middle East and North Africa

In [130]:
Happyness_in_world.Country.sort_values().unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina',
       'Armenia', 'Australia', 'Austria', 'Azerbaijan', 'Bahrain',
       'Bangladesh', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Canada', 'Central African Republic', 'Chad', 'Chile', 'China',
       'Colombia', 'Comoros', 'Congo (Brazzaville)', 'Congo (Kinshasa)',
       'Costa Rica', 'Croatia', 'Cyprus', 'Czech Republic', 'Denmark',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador', 'Estonia',
       'Ethiopia', 'Finland', 'France', 'Gabon', 'Gambia', 'Georgia',
       'Germany', 'Ghana', 'Greece', 'Guatemala', 'Guinea', 'Haiti',
       'Honduras', 'Hong Kong', 'Hungary', 'Iceland', 'India',
       'Indonesia', 'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy',
       'Ivory Coast', 'Jamaica', 'Japan', 'Jordan', 'Kazakhstan', 'Kenya',
       'Kosovo', 'Kuwait'

In [131]:
#store the final dataset into a .csv file
Happyness_in_world.to_csv('happy_regions.csv')

In [ ]:
#open the dataset modified with the missing Regions 

In [132]:
Happyness_in_world_def=pd.read_csv('/Users/nuria/Documents/datasciencecourse/FINAL PROJECT/happy_regions.csv')

In [133]:
#drop the column unnamed generated when we have created ths csv file
Happyness_in_world_def.drop(['Unnamed: 0'], axis=1,inplace=True)

In [134]:
Happyness_in_world_def.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 620 entries, 0 to 619
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Year                620 non-null    int64  
 1   Country             620 non-null    object 
 2   Rank                620 non-null    int64  
 3   Happiness_score     620 non-null    float64
 4   Live_expect_score   620 non-null    float64
 5   Helth_life_exp_age  620 non-null    float64
 6   Region              586 non-null    object 
dtypes: float64(3), int64(2), object(2)
memory usage: 34.0+ KB


In [135]:
Happyness_in_world.rename(columns={"Helth_life_exp_age": "Health_life_exp_age"},inplace=True)

In [136]:
Happyness_in_world

Year                   Country  Rank  Happiness_score  Live_expect_score  \
0    2017                    Norway     1           7.5370           0.796667   
1    2017                   Denmark     2           7.5220           0.792566   
2    2017                   Iceland     3           7.5040           0.833552   
3    2017               Switzerland     4           7.4940           0.858131   
4    2017                   Finland     5           7.4690           0.809158   
..    ...                       ...   ...              ...                ...   
615  2020  Central African Republic   149           3.4759           0.000000   
616  2020                    Rwanda   150           3.3123           0.572383   
617  2020                  Zimbabwe   151           3.2992           0.375038   
618  2020               South Sudan   152           2.8166           0.208809   
619  2020               Afghanistan   153           2.5669           0.266052   

     Health_life_exp_age              Region  
0              73.099998      Western Europe  
1              72.099998      Western Europe  
2              73.000000      Western Europe  
3              73.800003      Western Europe  
4              71.800003      Western Europe  
..                   ...                 ...  
615            45.200001  Sub-Saharan Africa  
616            61.098846  Sub-Saharan Africa  
617            55.617260  Sub-Saharan Africa  
618            51.000000  Sub-Saharan Africa  
619            52.590000          South Asia  

[620 rows x 7 columns]

In [137]:
# Store the dataset to use it for the EDA
Happyness_in_world_def.to_csv('Happyness_in_world_definitive.csv',index=False)